h3-quinn/src/lib.rs

```diff
@@ -0,0 +1,291 @@
+    ) -> Poll<Result<Option<Self::Buf>, Self::Error>> {
+        self.recv.poll_data(cx)
+    }
+
+    fn stop_sending(&mut self, error_code: u64) {
+        self.recv.stop_sending(error_code)
+    }
+}
+
+impl<B, S> quic::SendStream<B> for QuinnBidiStream<B, S>
+where
+    B: Buf,
+    S: Session,
+{
+    type Error = SendStreamError;
+
+    fn poll_ready(&mut self, cx: &mut task::Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.send.poll_ready(cx)
+    }
+
+    fn poll_finish(&mut self, cx: &mut task::Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.send.poll_finish(cx)
+    }
+
+    fn reset(&mut self, reset_code: u64) {
+        self.send.reset(reset_code)
+    }
+
+    fn send_data(&mut self, data: B) -> Result<(), Self::Error> {
+        self.send.send_data(data)
+    }
+}
+
+struct QuinnRecvStream<S: Session> {
+    stream: RecvStream<S>,
+    offset: u64,
+    chunks: BTreeMap<u64, Bytes>,
+}
+
+impl<S: Session> QuinnRecvStream<S> {
+    fn new(stream: RecvStream<S>) -> Self {
+        Self {
+            stream,
+            offset: 0,
+            chunks: BTreeMap::new(),
+        }
+    }
+}
+
+impl<S: Session> quic::RecvStream for QuinnRecvStream<S> {
+    type Buf = Bytes;
+    type Error = ReadError;
+
+    fn poll_data(
+        &mut self,
+        cx: &mut task::Context<'_>,
+    ) -> Poll<Result<Option<Self::Buf>, Self::Error>> {
+        let ret = match self.stream.read_unordered().poll_unpin(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Ok(None)) => Poll::Ready(Ok(None)),
+            Poll::Ready(Err(e)) => return Poll::Ready(Err(e.into())),
+            // If we get the chunk we're looking for, return it right away
+            Poll::Ready(Ok(Some((mut chunk, offset)))) if offset <= self.offset => {
+                chunk.advance((self.offset - offset) as usize); // XXX overflow
+                self.offset += chunk.len() as u64;
+                return Poll::Ready(Ok(Some(chunk)));
+            }
+            // A chunk beyond current offset gets saved
+            Poll::Ready(Ok(Some((data, offset)))) => {
+                self.chunks.insert(offset, data);
+                Poll::Pending
+            }
+        };
+
+        // Nothing we've read can be yeilded, but we could have some chunk corresponding to `offset`
+        let chunk_key = self
+            .chunks
+            .keys()
+            .take_while(|x| **x <= self.offset)
+            .next()
+            .map(|x| *x);
+        if let Some(offset) = chunk_key {
+            let mut chunk = self.chunks.remove(&offset).unwrap();
+            chunk.advance((self.offset - offset) as usize); // XXX overflow
+            self.offset += chunk.len() as u64;
+            return Poll::Ready(Ok(Some(chunk)));
+        };
+
+        ret
+    }
+
+    fn stop_sending(&mut self, error_code: u64) {
+        let _ = self
+            .stream
+            .stop(VarInt::from_u64(error_code).expect("invalid error_code"));
```

Comment:
Do you think it's fine to panic if the error code is invalid here ? Or could we transfer the bound check to the user, accepting something like `VarInt` ?

Reply:
Yea, this was something I couldn't decide. I think there's these options (I could be missing some):

- Add some `h3::quic::ApplicationErrorCode` type that enforces the invariants. The downside is you'd probably need to convert *that* into a `u64` and then convert that into the QUIC-library-specific type, which means just check the value twice...
- Pass a `u64` and if it is bad, panic.
- Pass a `u64` and if it is bad, ignore it.
- Pass a `u64`, and if it is bad, return an error. That'd need us to change `stop_sending` to return `Result<(), Self::Error>`.

Reply:
After all I guess setting it to `VarInt::MAX` if beyond the (sole) upper bound is a good trade-off. This is what I've just done. Normally, users should pass an `HttpError` enum value, and an unknown value does not represent a critical condition for the peer. 

This should be the broadest behavior, to fit most QUIC impls. 

Note that `VarInt` is not an only-QUIC thing. It is used for most integer values in h3 also: frame headers, stream types. And, it is a very simple type that as a upper bound check and is encoded with a "size hint" prefix on the most significant bits. It is not really an implementation-specific behavior.

`quinn-h3` uses `quinn::VarInt` and I think `h3` could as well. I'll open an issue for that matter.